### PR TITLE
implemented intersectWith and intersect functions

### PR DESCRIPTION
**Pull request by pnobre**
Implemented intersectWith and intersect functions on Maps, Dictionaries and Read-Only Dictionaries.

tests will follow once previous PR get's merged in order to avoid creating another file.

### CODE REVIEW

File: src/FSharpPlus/Extensions.fs [hunk 1]
```diff
@@ -377,7 +377,24 @@ module Map =
         Map.fold (fun m k v' -> Map.add k (match Map.tryFind k m with Some v -> combiner v v' | None -> v') m) source1 source2
 
     /// Returns the union of two maps, preferring values from the first in case of duplicate keys.
-    let union (source: Map<'Key, 'T>) (altSource: Map<'Key, 'T>) = unionWith (fun x _ -> x) source altSource
+    let union (source: Map<'Key, 'T>) (altSource: Map<'Key, 'T>) = unionWith (fun k _ -> k) source altSource
```
Comment (wallymathieu) [inline src/FSharpPlus/Extensions.fs]: `x`->`k`

**Comment by wallymathieu** [comment]
The build failure on Travis looks unrelated

**Comment by wallymathieu** [comment]
What's the status here? Is it OK? I've merged a fix for travis build failure, so this can be merged if it's OK.

**Comment by wallymathieu** [comment]
Ping @pnobre ?

**Comment by pnobre** [comment]
Hey @wallymathieu, IIRC this PR was missing some unit tests requested by @gusty . Have been quite busy latelly, will get back to it in the next couple of days to create the tests and update the PR.

**Comment by wallymathieu** [comment]
Just trying to understand the state of things, no stress

File: src/FSharpPlus/Extensions.fs (hunk 1 above)
Comment (pnobre) [reply src/FSharpPlus/Extensions.fs]: what do you mean `x -> k` @wallymathieu ? replace x with k ?

**Comment by gusty** [comment]
This looks good, thanks for your contribution !!!

**Pull request merged**